src/hazelcore/dynamics/Evaluator.re

```diff
@@ -528,9 +526,15 @@ let eval_bin_float_op =
   };
 };
 
-let rec evaluate = (d: DHExp.t): result =>
+let rec evaluate = (d: DHExp.t): EvaluatorResult.t =>
   switch (d) {
-  | BoundVar(_) => InvalidInput(1)
+  | BoundVar(x) =>
+    switch (Builtins.lookup_type(x)) {
+    | Some(ty) =>
+      // TODO: Assumes all builtins are functions
```

Comment:
this approach also means we can't shadow built-in functions, which I think would be fine to allow. 

I think a better approach is to insert all of the built-ins as let bindings wrapping ApBuiltin at the top of the program after elaboration is complete.

Reply:
I added a new `elab` function to `Elaborator_Exp.re` that we call from the outside instead of `syn_elab` directly. In there I insert the builtins at the top. Is that what you were thinking?